sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -123,12 +124,30 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
   /**
    * :: Experimental ::
    * Specifies the underlying output data source. Built-in options include "parquet", "json", etc.
+   * Additionally user specific StreamSinkProviders can be specified hear using the fully qualified
+   * class name.
    *
    * @since 2.0.0
    */
   @Experimental
   def format(source: String): DataStreamWriter[T] = {
     this.source = source
+    this.sinkProvider = null
+    this
+  }
+
+  /**
+   * :: Experimental ::
+   * Specifies the underlying output data source using a StreamSinkProvider. This is useful for
+   * sinks which are constructed with user specified functions (such as a user specified version of
+   * ForeachSink).
+   *
+   * @since 2.1.0
+   */
+  @Experimental
+  def format(sinkProvider: StreamSinkProvider): DataStreamWriter[T] = {
+    this.source = null
```

Comment:
Probably better to set `source` to a dedicated constant string like "custom" here, so that the branches in start() all have consistent logic.